### PR TITLE
Invalid tx should not be included in block

### PR DESCRIPTION
**Pull request by bvbfan**
/kind fix

#### What this PR does / why we need it:

<s>It has 2 issues about me:</s>

1. Failed tx never be part of a block, i try to ensure it
<s>2. We use a view to ensure txs validation in a block not in isolation, but we fail to discard invalid view.</s>

### CODE REVIEW

**Comment by Bushstar** [comment]
Note that it is native Bitcoin behaviour to have `failedTx.insert(iter)` in a `fUsingModified` conditional. 

https://github.com/bitcoin/bitcoin/blob/0.19/src/miner.cpp#L373

File: src/txmempool.cpp [hunk 1]
```diff
@@ -594,10 +594,13 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef>& vtx, unsigne
...
                 continue;
             }
-            auto res = ApplyCustomTx(viewDuplicate, mempoolDuplicate, tx, Params().GetConsensus(), nBlockHeight, 0, false);
+            CCustomCSView workView(viewDuplicate); // we don't need invalid tx to change the view
+            auto res = ApplyCustomTx(workView, mempoolDuplicate, tx, Params().GetConsensus(), nBlockHeight, 0, false);
             if (!res.ok && (res.code & CustomTxErrCodes::Fatal)) {
                 LogPrintf("%s: Remove conflicting custom TX: %s\n", __func__, tx.GetHash().GetHex());
                 staged.insert(mapTx.project<0>(it));
+            } else {
+                workView.Flush(); // apply changes when they are checked
```
Comment (ShengguangXiao) [inline src/txmempool.cpp:603]: Inside `ApplyCustomTx` if it is success, it already flushed, why here need to flush again?

Comment (bvbfan) [reply src/txmempool.cpp:603]: Yeah, i forgot that `ApplyCustomTx` do it for us, when tx is invalid it does not flush.

**Comment by Bushstar** [comment]
The last entry is bespoke to us and 

```
        // Failed, let's move on!
        if (!customTxPassed) {
            if (fUsingModified) {
                mapModifiedTx.get<ancestor_score>().erase(modit);
            }
            failedTx.insert(iter);
            continue;
        }
```

**Review by ShengguangXiao — APPROVED** [review]

**Pull request merged**